File: app/src/main/java/com/mancode/financetracker/ui/transactions/FilterQuery.java
package com.mancode.financetracker.ui.transactions;

import com.mancode.financetracker.database.converter.DateConverter;
import com.mancode.financetracker.database.pojos.TransactionFull;

import org.threeten.bp.LocalDate;

public class FilterQuery {

    private static final int TOKENS_COUNT = 5;

    private static final int TOKEN_TYPE = 0;
    private static final int TOKEN_FROM_DATE = 1;
    private static final int TOKEN_TO_DATE = 2;
    private static final int TOKEN_TIMESPAN = 3;
    private static final int TOKEN_BOOKMARK = 4;

    static final int TYPE_ALL = 0;
    static final int TYPE_INCOME = 1;
    static final int TYPE_OUTCOME = -1;

    static final int UNCONSTRAINED = 0;
    static final int LAST_WEEK = 1;
    static final int LAST_MONTH = 2;
    static final int THIS_MONTH = 3;
    static final int PREVIOUS_MONTH = 4;
    static final int THIS_YEAR = 5;
    static final int CUSTOM = 6;

    private String mQuery;
    private String[] mTokens;
    private static final String SEPARATOR = ",";

    FilterQuery(String query) {
        this.mQuery = query;
        tokenize();
    }

    FilterQuery(Integer type, LocalDate from, LocalDate to, Integer timespan, boolean bookmark) {
        mQuery = type +
                SEPARATOR +
                DateConverter.toString(from) +
                SEPARATOR +
                DateConverter.toString(to) +
                SEPARATOR +
                timespan +
                SEPARATOR +
                bookmark;
        tokenize();
    }

    static String getEmptyQuery() {
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < TOKENS_COUNT - 1; i++) {
            sb.append(SEPARATOR);
        }
        return sb.toString();
    }

    private void tokenize() {
        mTokens = mQuery.split(SEPARATOR, -1);
    }

    public String getQuery() {
        return mQuery;
    }

    public int getType() {
        String token = mTokens[TOKEN_TYPE];
        return !token.isEmpty() ? Integer.parseInt(token) : TYPE_ALL;
    }

    private LocalDate getFromDate() {
        return DateConverter.toDate(mTokens[TOKEN_FROM_DATE]);
    }

    private LocalDate getToDate() {
        return DateConverter.toDate(mTokens[TOKEN_TO_DATE]);
    }

    public int getTimespan() {
        String token = mTokens[TOKEN_TIMESPAN];
        return !token.isEmpty() ? Integer.parseInt(token) : UNCONSTRAINED;
    }

    public boolean bookmarked() {
        String token = mTokens[TOKEN_BOOKMARK];
        return !token.isEmpty() && Boolean.parseBoolean(token);
    }

    boolean isMatch(TransactionFull transaction) {
        int type = getType();
        LocalDate fromDate = getFromDate();
        LocalDate toDate = getToDate();

        return (type == TYPE_ALL || type == transaction.getTransaction().getType()) &&
                (fromDate == null || !transaction.getTransaction().getDate().isBefore(fromDate)) &&
                (toDate == null || !transaction.getTransaction().getDate().isAfter(toDate)) &&
                (!bookmarked() || transaction.getTransaction().getFlags() == 1);
    }
}

File: app/src/main/java/com/mancode/financetracker/ui/transactions/FilterQuery.kt
package com.mancode.financetracker.ui.transactions

import com.mancode.financetracker.database.pojos.TransactionFull
import org.threeten.bp.LocalDate
import java.util.Locale.ROOT

class FilterQuery {
    var description: String = ""
        private set(value) {
            field = value.trim().replace(SEPARATOR, "")
        }
    var type: Int = TYPE_ALL
        private set
    var fromDate: LocalDate? = null
        private set
    var toDate: LocalDate? = null
        private set
    var timespan: Int = UNCONSTRAINED
        private set
    var bookmarked: Boolean = false
        private set

    fun isEmpty(): Boolean =
            description.isEmpty()
                    && type == TYPE_ALL
                    && fromDate == null
                    && toDate == null
                    && timespan == UNCONSTRAINED
                    && !bookmarked

    fun update(
            description: String? = null,
            type: Int? = null,
            from: LocalDate? = null,
            to: LocalDate? = null,
            timespan: Int? = null,
            bookmark: Boolean? = null
    ) {
        if (description != null) this.description = description
        if (type != null) this.type = type
        if (from != null) this.fromDate = from
        if (to != null) this.toDate = to
        if (timespan != null) {
            this.timespan = timespan
            if (timespan == UNCONSTRAINED) {
                this.fromDate = null
                this.toDate = null
            }
        }
        if (bookmark != null) this.bookmarked = bookmark
    }

    fun resetExceptDescription() {
        update(
                type = TYPE_ALL,
                timespan = UNCONSTRAINED,
                bookmark = false
        )
    }

    fun isMatch(transaction: TransactionFull): Boolean {
        return (description.isEmpty() || transaction.transaction.description
                .toLowerCase(ROOT).contains(description.toLowerCase(ROOT))) &&
                (type == TYPE_ALL || type == transaction.transaction.type) &&
                (fromDate == null || !transaction.transaction.date.isBefore(fromDate)) &&
                (toDate == null || !transaction.transaction.date.isAfter(toDate)) &&
                (!bookmarked || transaction.transaction.flags == 1)
    }

    companion object {
        const val TYPE_ALL = 0
        const val TYPE_INCOME = 1
        const val TYPE_OUTCOME = -1

        const val UNCONSTRAINED = 0
        const val LAST_WEEK = 1
        const val LAST_MONTH = 2
        const val THIS_MONTH = 3
        const val PREVIOUS_MONTH = 4
        const val THIS_YEAR = 5
        const val CUSTOM = 6

        private const val SEPARATOR = ","
    }
}